### PR TITLE
Whois Rate Limit Handling

### PR DESCRIPTION
**Pull request by tkatzir**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/XSUP-24137

### CODE REVIEW

File: Packs/Whois/ReleaseNotes/1_4_15.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#### Integrations
+##### Whois
+- Added support for Rate Limit handling with **rate_limit_retry_count** instance configuration field, and a command param. Default value: 3 times.
```
Comment (ShirleyDenkberg) [inline Packs/Whois/ReleaseNotes/1_4_15.md:3]: Combine the 3 lines as follows:
- Added support for Rate Limit handling with the following instance configuration fields and command parameters:
   - *rate_limit_retry_count* - Default: 3 times.
   - *rate_limit_wait_seconds* - Default: 120 seconds.
   - *rate_limit_errors_suppressed* - Default: False.

**Comment by ShirleyDenkberg** [comment]
@tkatzir Doc review completed.

File: Packs/Whois/Integrations/Whois/Whois.py [hunk 2]
```diff
@@ -10,6 +10,10 @@
 
 SHOULD_ERROR = demisto.params().get('with_error', False)
 
+RATE_LIMIT_RETRY_COUNT_DEFAULT: int = 3
```
Comment (anara123) [inline Packs/Whois/Integrations/Whois/Whois.py:13]: Add another note under README -> Troubleshooting of the integration where you explain if the !ip command finishes with timeout, possible reason could be exceeding rate limit. And possible solutions will be either investigate why they reached exceeding rate limit, or set the retry to 0. This way the commands not timeout, but fail immediately. 
image

**Review by anara123 — CHANGES_REQUESTED** [review]

File: Packs/Whois/Integrations/Whois/Whois.py [hunk 3]
```diff
@@ -8591,7 +8609,18 @@ def main():
 
     try:
         if command == 'ip':
-            return_results(ip_command(demisto.args().get('ip'), reliability))
+            demisto_args = demisto.args()
```
[the remark this answers is not in the document]
Comment (tkatzir) [reply Packs/Whois/Integrations/Whois/Whois.py:8617]: https://github.com/demisto/content/pull/26579/commits/403c9570a6d7ede6cfd4b077bc62e4f155c36af3

File: Packs/Whois/Integrations/Whois/Whois.py [hunk 4]
```diff
@@ -8591,7 +8609,18 @@ def main():
...
+            ip = demisto_args.get('ip')
+            demisto_params = demisto.params()
+            rate_limit_retry_count: int = int(demisto_params.get('rate_limit_retry_count') or demisto_args.get('rate_limit_retry_count') or RATE_LIMIT_RETRY_COUNT_DEFAULT)
```
[the remark this answers is not in the document]
Comment (tkatzir) [reply Packs/Whois/Integrations/Whois/Whois.py]: https://github.com/demisto/content/pull/26579/commits/403c9570a6d7ede6cfd4b077bc62e4f155c36af3

File: Packs/Whois/README.md [hunk 5]
```diff
@@ -8,6 +8,16 @@ The playbooks included in this pack help you save time and automate repetitive t
 Automating the process of getting domain registrant information and verifying if the registrant is part of the user’s approved list.
 Automating the process of checking the domain registration age according to the predefined minimum age provided by the user.
 
+##### Troubleshooting
+
+###### `ip` command timeout
+The `ip` command can timeout when there are exssesive calls to the whois service, which will rate limits these calls.
+In that case, the default behavior is to try 3 times, with a 120 seconds wait between retries.
```
Comment (ShirleyDenkberg) [inline Packs/Whois/README.md:15]:
```suggestion
In that case, the default behavior is to try 3 times, with a 120 second wait between retries.
```

File: Packs/Whois/Integrations/Whois/Whois.py (hunk 2 above)
Comment (tkatzir) [reply Packs/Whois/Integrations/Whois/Whois.py:13]: https://github.com/demisto/content/pull/26579/commits/3aefe389ea89693fe7dce2b5bc9ea9c1a3626605

**Comment by tkatzir** [comment]
DBotPredictURLPhishing_test failed, but it succeeded on the test machine when I ran it.

**Comment by tkatzir** [comment]
Force-merge due to TPB failing, unrelated to this PR.

**Pull request merged**